File: yntra-ios/YntraIOS/TodoViewModel.swift
import SwiftUI
import yntra_core

@MainActor
class TodoViewModel: ObservableObject {
    @Published var todos: [TodoItem] = []
    @Published var errorMessage: String? = nil
    private var observer: SwiftDbObserver?
    
    init() {
        // Initial Load
        loadTodos()
        
        // Setup observer callback
        self.observer = SwiftDbObserver { [weak self] in
            self?.loadTodos()
        }
        
        // Register observer to Rust FFI boundary
        registerObserver(observer: self.observer!)
    }
    
    func loadTodos() {
        Task {
            do {
                self.todos = try await getTodos(requesterUserId: "user-1", workspaceId: "workspace-1")
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }
    
    func addNewTodo(text: String) {
        Task {
            do {
                _ = try await addTodo(requesterUserId: "user-1", workspaceId: "workspace-1", text: text)
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }
    
    func toggle(todo: TodoItem) {
        Task {
            do {
                try await toggleTodo(requesterUserId: "user-1", id: todo.id)
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }
}

File: yntra-android/app/src/main/java/com/yntra/app/TodoViewModel.kt
package com.yntra.app

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import uniffi.yntra_core.*

class TodoViewModel : ViewModel() {
    private val _todos = MutableStateFlow<List<TodoItem>>(emptyList())
    val todos: StateFlow<List<TodoItem>> = _todos
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage
    
    private val observer = KotlinDbObserver {
        loadTodos()
    }
    
    init {
        loadTodos()
        // Register observer to Rust FFI boundary
        registerObserver(observer)
    }
    
    private fun loadTodos() {
        viewModelScope.launch {
            try {
                // Pass dynamic user and workspace IDs to FFI function
                _todos.value = getTodos(SessionManager.activeUserId, SessionManager.activeWorkspaceId)
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }
    
    fun addNewTodo(text: String) {
        viewModelScope.launch {
            try {
                addTodo(SessionManager.activeUserId, SessionManager.activeWorkspaceId, text)
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }

    fun toggle(todo: TodoItem) {
        viewModelScope.launch {
            try {
                toggleTodo(SessionManager.activeUserId, todo.id)
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }
    
    override fun onCleared() {
        super.onCleared()
        // Prevent FFI callback memory leaks
        clearObservers()
    }
}
